org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -937,6 +946,37 @@ private int putData(byte[] data, ByteBuffer[] buffers, int offset, int length) {
         return data_index;
     }
 
+    private void updateSession() {
+        if (ssl_fd == null) {
+            return;
+        }
+
+        try {
+            PK11Cert[] peer_chain = SSL.PeerCertificateChain(ssl_fd);
+            session.setPeerCertificates(peer_chain);
+
+            SSLChannelInfo info = SSL.GetChannelInfo(ssl_fd);
+            if (info == null) {
+                return;
+            }
+
+            session.setId(info.getSessionID());
+            session.refreshData();
+        
+            NextProtoResult ret = SSL.GetNextProto(ssl_fd);
+            if (ret != null) {
+                // Only advertise the resulting protocol if we have one.
+                if (ret.state != SSLNextProtoState.SSL_NEXT_PROTO_NO_SUPPORT &&
+                    ret.state != SSLNextProtoState.SSL_NEXT_PROTO_NO_OVERLAP)
```

Comment:
We should be more defensive here; also check that `ret.getProtocol()` is both non-null and non-empty.

Reply:
@cipherboy I just noticed that the return contract of `getApplicationProtocol` (is SSLEngine and SSLSocket and probably other places) is:

> null if it has not yet been determined if application protocols might be used for this connection, an empty String if application protocols values will not be used, or a non-empty application protocol String if a value was successfully negotiated.

So it seems that we should always `setNextProtocol`

- empty string for `NO_SUPPORT` and `NO_OVERLAP`
- `ret.getProtocol()` for other cases.